From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: 404Setup <153366651+404Setup@users.noreply.github.com>
Date: Wed, 29 May 2024 21:59:04 +0800
Subject: [PATCH] Fix-build

and add kotlin support

diff --git a/build.gradle.kts b/build.gradle.kts
index c2d3d699edfd60d773af96116c5663c812c691e9..524404ce77a6f48f35bd618b020db6bb9ad0587d 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -4,6 +4,9 @@ import java.time.Instant
 plugins {
     java
     `maven-publish`
+    kotlin("jvm") version "2.0.0"
+    kotlin("plugin.serialization") version "2.0.0"
+
 }
 
 val log4jPlugins = sourceSets.create("log4jPlugins")
@@ -13,7 +16,7 @@ configurations.named(log4jPlugins.compileClasspathConfigurationName) {
 val alsoShade: Configuration by configurations.creating
 
 dependencies {
-    implementation(project(":luminol-api")) // Folia // Luminol
+    implementation(project(":levelbukkit-api")) // Folia // Luminol // LevelBukkit
     implementation("com.electronwill.night-config:toml:3.6.6") // Luminol - Night config
     // Paper start
     implementation("org.jline:jline-terminal-jansi:3.21.0")
@@ -70,6 +73,14 @@ dependencies {
     implementation("me.lucko:spark-api:0.1-20240720.200737-2")
     implementation("me.lucko:spark-paper:1.10.84-20240720.204128-1")
     // Paper end - spark
+
+    // LevelBukkit start
+    implementation(kotlin("reflect"))
+    implementation("org.jetbrains.kotlinx:kotlinx-coroutines-core:1.9.0-RC")
+    implementation("org.jetbrains.kotlinx:kotlinx-io-core:0.4.0")
+    implementation("org.jetbrains.kotlinx:atomicfu:0.24.0")
+    implementation("org.jetbrains.kotlinx:kotlinx-serialization-json:1.7.0")
+    // LevelBukkit end
 }
 
 paperweight {
@@ -89,14 +100,14 @@ tasks.jar {
         val gitBranch = git("rev-parse", "--abbrev-ref", "HEAD").getText().trim() // Paper
         attributes(
             "Main-Class" to "org.bukkit.craftbukkit.Main",
-            "Implementation-Title" to "Luminol", // Folia // Luminol
+            "Implementation-Title" to "LevelBukkit", // Folia // Luminol // LevelBukkit
             "Implementation-Version" to implementationVersion,
             "Implementation-Vendor" to date, // Paper
-            "Specification-Title" to "Luminol", // Folia // Luminol
+            "Specification-Title" to "LevelBukkit", // Folia // Luminol // LevelBukkit
             "Specification-Version" to project.version,
-            "Specification-Vendor" to "LuminolMC Team", // Luminol
-            "Brand-Id" to "luminolmc:luminol", // Folia // Luminol
-            "Brand-Name" to "Luminol", // Folia // Luminol
+            "Specification-Vendor" to "TranicSoft LevelPowered", // Luminol // LevelBukkit
+            "Brand-Id" to "levelpowered:bukkit", // Folia // Luminol // LevelBukkit
+            "Brand-Name" to "levelbukkit", // Folia // Luminol // LevelBukkit
             "Build-Number" to (build ?: ""),
             "Build-Time" to Instant.now().toString(),
             "Git-Branch" to gitBranch, // Paper
@@ -109,6 +120,17 @@ tasks.jar {
     }
 }
 
+// LevelBukkit start
+kotlin {
+    compilerOptions {
+        languageVersion.set(org.jetbrains.kotlin.gradle.dsl.KotlinVersion.KOTLIN_2_0)
+        apiVersion.set(org.jetbrains.kotlin.gradle.dsl.KotlinVersion.KOTLIN_2_0)
+        jvmTarget.set(org.jetbrains.kotlin.gradle.dsl.JvmTarget.JVM_21)
+    }
+}
+// LevelBukkit end
+
+
 // Paper start - compile tests with -parameters for better junit parameterized test names
 tasks.compileTestJava {
     options.compilerArgs.add("-parameters")
