From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: 404Setup <153366651+404Setup@users.noreply.github.com>
Date: Wed, 29 May 2024 23:02:26 +0800
Subject: [PATCH] Divine: Fix MC-172801


diff --git a/src/main/java/net/minecraft/world/entity/LivingEntity.java b/src/main/java/net/minecraft/world/entity/LivingEntity.java
index 7aade876648e79e48b80b4048ba0f8c07483bd96..aee6ae8f93b53da1c1800e886d5722ca92fc72be 100644
--- a/src/main/java/net/minecraft/world/entity/LivingEntity.java
+++ b/src/main/java/net/minecraft/world/entity/LivingEntity.java
@@ -3181,7 +3181,13 @@ public abstract class LivingEntity extends Entity implements Attackable {
     }
 
     protected float getFlyingSpeed() {
-        return this.getControllingPassenger() instanceof net.minecraft.world.entity.player.Player ? this.getSpeed() * 0.1F : 0.02F;
+        // DivineMC start - Fix MC-172801
+        float flyingSpeed = 0.02F;
+        if (this.getAttributes().hasAttribute(Attributes.FLYING_SPEED)) {
+            flyingSpeed = (float) (this.getAttribute(Attributes.FLYING_SPEED).getValue() * 0.049999999254942D);
+        }
+        return this.getControllingPassenger() instanceof net.minecraft.world.entity.player.Player ? this.getSpeed() * 0.1F : flyingSpeed;
+        // DivineMC end
     }
 
     public float getSpeed() {
