From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: 404Setup <153366651+404Setup@users.noreply.github.com>
Date: Fri, 31 May 2024 11:24:45 +0800
Subject: [PATCH] Pufferfish: Add SIMD utilities


diff --git a/build.gradle.kts b/build.gradle.kts
index 4d74dc28565cbe42fbc1ae4348c90c7a7c4a2496..86bf2b8373fed052b2db1d6d7ab5b15858fa2829 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -79,6 +79,13 @@ paperweight {
     craftBukkitPackageVersion.set("v1_20_R4") // also needs to be updated in MappingEnvironment
 }
 
+// Pufferfish Start
+tasks.withType<JavaCompile> {
+    val compilerArgs = options.compilerArgs
+    compilerArgs.add("--add-modules=jdk.incubator.vector")
+}
+// Pufferfish End
+
 tasks.jar {
     archiveClassifier.set("dev")
 
