From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: 404Setup <153366651+404Setup@users.noreply.github.com>
Date: Fri, 31 May 2024 11:24:45 +0800
Subject: [PATCH] Pufferfish: Add SIMD utilities


diff --git a/build.gradle.kts b/build.gradle.kts
index 14b9e2d8a87a5d2478ea83c8fb3313df834f152e..0f9c40de2c3e38429cdfc0d7c787c36c9dae352c 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -78,6 +78,14 @@ dependencies {
 }
 
 val craftbukkitPackageVersion = "1_20_R3" // Paper
+
+// Pufferfish Start
+tasks.withType<JavaCompile> {
+    val compilerArgs = options.compilerArgs
+    compilerArgs.add("--add-modules=jdk.incubator.vector")
+}
+// Pufferfish End
+
 tasks.jar {
     archiveClassifier.set("dev")
 
