From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: 404Setup <153366651+404Setup@users.noreply.github.com>
Date: Fri, 31 May 2024 20:07:56 +0800
Subject: [PATCH] Purpur: Allow soil to moisten from water directly under it


diff --git a/src/main/java/net/minecraft/world/level/block/FarmBlock.java b/src/main/java/net/minecraft/world/level/block/FarmBlock.java
index edb22983a17cc22ba01ee28a9ebc685aaae117eb..455fdd1714972229534915aa3619a37aac0228b3 100644
--- a/src/main/java/net/minecraft/world/level/block/FarmBlock.java
+++ b/src/main/java/net/minecraft/world/level/block/FarmBlock.java
@@ -30,6 +30,7 @@ import net.minecraft.world.level.pathfinder.PathComputationType;
 import net.minecraft.world.phys.shapes.CollisionContext;
 import net.minecraft.world.phys.shapes.VoxelShape;
 // CraftBukkit start
+import one.tranic.bukkit.config.LevelBukkitConfig;
 import org.bukkit.event.entity.EntityInteractEvent;
 import org.bukkit.craftbukkit.event.CraftEventFactory;
 // CraftBukkit end
@@ -172,7 +173,7 @@ public class FarmBlock extends Block {
             }
         }
 
-        return false;
+        return (LevelBukkitConfig.Feature.farmlandGetsMoistFromBelow && world.getFluidState(pos.relative(Direction.DOWN)).is(FluidTags.WATER)); // Purpur;
         // Paper end - Perf: remove abstract block iteration
     }
 
