From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: 404Setup <153366651+404Setup@users.noreply.github.com>
Date: Fri, 31 May 2024 20:07:56 +0800
Subject: [PATCH] Purpur: Allow soil to moisten from water directly under it


diff --git a/src/main/java/net/minecraft/world/level/block/FarmBlock.java b/src/main/java/net/minecraft/world/level/block/FarmBlock.java
index ab2b9b85b844cf3e6ef5a0379fa3279b85a7328c..8dfb7342d7ae603114613f1c7310d245b8121900 100644
--- a/src/main/java/net/minecraft/world/level/block/FarmBlock.java
+++ b/src/main/java/net/minecraft/world/level/block/FarmBlock.java
@@ -3,6 +3,8 @@ package net.minecraft.world.level.block;
 import com.mojang.serialization.MapCodec;
 import java.util.Iterator;
 import javax.annotation.Nullable;
+
+import dev.paged.levelbukkit.config.LevelBukkitConfig;
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.Direction;
 import net.minecraft.server.level.ServerLevel;
@@ -171,7 +173,8 @@ public class FarmBlock extends Block {
             }
         }
 
-        return false;
+
+        return (LevelBukkitConfig.Purpur.World.farmlandGetsMoistFromBelow && world.getFluidState(pos.relative(Direction.DOWN)).is(FluidTags.WATER)); // Purpur;
         // Paper end - Perf: remove abstract block iteration
     }
 
